### PR TITLE
add HSIC metric

### PR DESCRIPTION
**Pull request by kzkadc**
Description: added Hilbert-Schmidt Independence Criterion (HSIC).

HSIC is a kernel-based method to measure the statistical independence between two feature vectors:

https://papers.nips.cc/paper_files/paper/2007/hash/d5cfead94f5350c12c322b5b664544c1-Abstract.html
https://jejjohnson.github.io/research_journal/appendix/similarity/hsic/

This metric computes the unbiased estimator of the HSIC proposed in the following:

https://jmlr.csail.mit.edu/papers/v13/song12a.html

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/hsic.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            Y = torch.sin(X * torch.pi * 2 / 50)
+            state = default_evaluator.run([[X, Y]])
+            print(state.metrics["hsic"])
+
+        .. testoutput::
+
+            0.0922664999961853
+
+    .. versionchanged:: 0.5.1
+        ``skip_unrolling`` argument is added.
```
Comment (vfdev-5) [inline ignite/metrics/hsic.py:92]: This should replaced by 
```python
.. versionadded:: 0.5.2
```
(link: https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#writing-documentation)


By the way, for the previous PR with Kendall and Spearman correlation metrics, can you please also send a PR with `versionadded` tag? 
We may also want to improve the docstring and provide the docs for missing ctor args (`skip_unrolling` and `check_compute_fn`), see for example: https://deploy-preview-3282--pytorch-ignite-preview.netlify.app/generated/ignite.metrics.regression.spearmanrankcorrelation#spearmanrankcorrelation

File: ignite/metrics/hsic.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+        self._num_batches += 1
+
+    @sync_all_reduce("_sum_of_trace", "_sum_of_second_term", "_sum_of_third_term", "_num_batches")
+    def compute(self) -> float:
+        if self._num_batches == 0:
+            raise NotComputableError("HSIC must have at least one batch before it can be computed.")
+
+        hsic = self._sum_of_trace + self._sum_of_second_term - self._sum_of_third_term * 2.0
+        hsic = hsic.clamp(min=0.0)  # HSIC must not be negative
```
Comment (vfdev-5) [inline ignite/metrics/hsic.py:168]: I wonder whether in the paper they say somewhere to clip the value ?

File: tests/ignite/metrics/test_hsic.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    dyy = np.square(y_np[ii] - y_np[jj]).sum(axis=2)
+    if sigma_y < 0:
+        vy = np.median(dyy)
+    else:
+        vy = sigma_y * sigma_y
+    L = np.exp(-0.5 * dyy / vy) * mask
+
+    KL = K @ L
+    ones = np.ones(b)
+    hsic = np.trace(KL) + (ones @ K @ ones) * (ones @ L @ ones) / ((b - 1) * (b - 2)) - ones @ KL @ ones * 2 / (b - 2)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_hsic.py:39]: Same above question, we do not clip the value in the expected HSIC computation formula ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot for the PR @kzkadc !
I have few questions and comments to address.

File: ignite/metrics/hsic.py (hunk 2 above)
Comment (kzkadc) [reply ignite/metrics/hsic.py:168]: Thanks for your review!

Although they do not use clipping, I observed the computation sometimes produced slight negative values because of numerical errors.
So, I added clipping because the HSIC is the norm of the covariance in definition (Eq. (3) of Song, et al. (2012)), and it must not be negative.

**Comment by kzkadc** [comment]
@vfdev-5 
Are there still problems with this PR?
If so, could you please let me know how I can resolve them?

File: tests/ignite/metrics/test_hsic.py (hunk 3 above)
Comment (vfdev-5) [reply tests/ignite/metrics/test_hsic.py:39]: @kzkadc should we clip in this expected value formula?

**Comment by vfdev-5** [comment]
@kzkadc please check mypy failure:
```
+ mypy --config-file mypy.ini
ignite/metrics/hsic.py:139: error: Incompatible types in assignment (expression
has type "float", variable has type "Tensor")  [assignment]
                vx = self.sigma_x**2
                     ^~~~~~~~~~~~~~~
ignite/metrics/hsic.py:148: error: Incompatible types in assignment (expression
has type "float", variable has type "Tensor")  [assignment]
                vy = self.sigma_y**2
                     ^~~~~~~~~~~~~~~
Found 2 errors in 1 file (checked 146 source files)
```

File: tests/ignite/metrics/test_hsic.py (hunk 3 above)
Comment (kzkadc) [reply tests/ignite/metrics/test_hsic.py:39]: I added clipping here and modified HSIC to clip value for each batch (https://github.com/pytorch/ignite/pull/3282/commits/43bd324578dce00f3161d867339aac50a6ad0a5a).

File: tests/ignite/metrics/test_hsic.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+        rank = idist.get_rank()
+        torch.manual_seed(12 + rank)
+
+        device = idist.device()
+        metric_devices = [torch.device("cpu")]
+        if device.type != "xla":
+            metric_devices.append(device)
+
+        lin = nn.Linear(n_dims_x, n_dims_y)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_hsic.py:140]: This change looks a bit strange to me to fix failing CI. Why do we need to put it here?

**Comment by vfdev-5** [comment]
@kzkadc some distributed tests are failing:
```python
=========================== short test summary info ============================
FAILED tests/ignite/metrics/test_hsic.py::TestDistributed::test_integration[gloo_cpu--1.0-1.0] - assert 3.780633494340978e-05 ± 2.0e-05 == 2.1274586519148087e-31
  
  comparison failed
  Obtained: 2.1274586519148087e-31
  Expected: 3.780633494340978e-05 ± 2.0e-05
FAILED tests/ignite/metrics/test_hsic.py::TestDistributed::test_integration[gloo_cpu-1.0--1.0] - assert 3.780633494340978e-05 ± 2.0e-05 == 5.500369049116216e-11
  
  comparison failed
  Obtained: 5.500369049116216e-11
  Expected: 3.780633494340978e-05 ± 2.0e-05
FAILED tests/ignite/metrics/test_hsic.py::TestDistributed::test_integration[gloo_cpu-1.0-1.0] - assert 3.780633494340978e-05 ± 2.0e-05 == 3.3436754842126586e-39
  
  comparison failed
  Obtained: 3.3436754842126586e-39
  Expected: 3.780633494340978e-05 ± 2.0e-05
========== 3 failed, 2 skipped, 3133 deselected, 2 warnings in 7.72s ===========
```
I haven't checked in details why, but from the first glance something looks weird in the output values which are almost zeros and expected is just close to zero. Any ideas why this happens?

File: tests/ignite/metrics/test_hsic.py (hunk 4 above)
Comment (kzkadc) [reply tests/ignite/metrics/test_hsic.py:140]: Thanks for checking.
It's my misunderstanding and I will revert the change.

**Comment by kzkadc** [comment]
I'm struggling with the numerical errors and have been checking the code, but I'm not sure why it happens for now.
I will continue to check the code.

File: tests/ignite/metrics/test_hsic.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            engine.run(data=data, max_epochs=1)
+
+            assert "hsic" in engine.state.metrics
+            res = engine.state.metrics["hsic"]
+
+            np_res = 0.0
+            for i in range(n_iters):
+                x_batch, y_batch = data_loader(i)
+                np_res += np_hsic(x_batch, y_batch)
+            np_res = np_res / n_iters
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_hsic.py:164]: I think this expected value computation is not correct. `test_integration` function in distributed tests is run by multiple processes, so `HSIC` will compute average HSIC over all processes. On the other hand, `np_res` is computed per process and thus wont match.

File: tests/ignite/metrics/test_hsic.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+            data = list(range(n_iters))
+            engine.run(data=data, max_epochs=1)
+
+            assert "hsic" in engine.state.metrics
+            res = engine.state.metrics["hsic"]
+
+            np_res = 0.0
+            for i in range(n_iters):
+                x_batch, y_batch = data_loader(i)
+                np_res += np_hsic(x_batch, y_batch)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_hsic.py:163]: There are also missing sigma_x, sigma_y input parameters into `np_hsic` function.

File: tests/ignite/metrics/test_hsic.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    np_hsic_sum = 0.0
+    n_iters = y.shape[0] // batch_size
+    for i in range(n_iters):
+        idx = i * batch_size
+        x_batch = x[idx : idx + batch_size]
+        y_batch = y[idx : idx + batch_size]
+
+        hsic.update((x_batch, y_batch))
+        np_hsic_sum += np_hsic(x_batch, y_batch, sigma_x, sigma_y)
+    np_res = np_hsic_sum / n_iters
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_hsic.py:105]: let's call this variable as `expected_hsic`

File: tests/ignite/metrics/test_hsic.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+            m = HSIC(sigma_x=sigma_x, sigma_y=sigma_y, device=metric_device)
+            m.attach(engine, "hsic")
+
+            data = list(range(n_iters))
+            engine.run(data=data, max_epochs=1)
+
+            assert "hsic" in engine.state.metrics
+            res = engine.state.metrics["hsic"]
+
+            np_res = 0.0
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_hsic.py:160]: Same here, let's call this as `expected_hsic`

File: tests/ignite/metrics/test_hsic.py (hunk 5 above)
Comment (vfdev-5) [reply tests/ignite/metrics/test_hsic.py:164]: Here is working test code:
```python
    @pytest.mark.parametrize("sigma_x", [-1.0, 1.0])
    @pytest.mark.parametrize("sigma_y", [-1.0, 1.0])
    def test_integration(self, sigma_x: float, sigma_y: float):
        tol = 2e-5
        n_iters = 100
        batch_size = 20
        n_dims_x = 100
        n_dims_y = 50

        rank = idist.get_rank()
        torch.manual_seed(12 + rank)

        device = idist.device()
        metric_devices = [torch.device("cpu")]
        if device.type != "xla":
            metric_devices.append(device)

        for metric_device in metric_devices:
            x = torch.randn((n_iters * batch_size, n_dims_x)).float().to(device)

            lin = nn.Linear(n_dims_x, n_dims_y).to(device)
            y = torch.sin(lin(x) * 100) + torch.randn(n_iters * batch_size, n_dims_y) * 1e-4

            def data_loader(i, input_x, input_y):
                return input_x[i * batch_size : (i + 1) * batch_size], input_y[i * batch_size : (i + 1) * batch_size]

            engine = Engine(lambda e, i: data_loader(i, x, y))

            m = HSIC(sigma_x=sigma_x, sigma_y=sigma_y, device=metric_device)
            m.attach(engine, "hsic")

            data = list(range(n_iters))
            engine.run(data=data, max_epochs=1)

            assert "hsic" in engine.state.metrics
            res = engine.state.metrics["hsic"]

            x = idist.all_gather(x)
            y = idist.all_gather(y)
            total_n_iters = idist.all_reduce(n_iters)

            np_res = 0.0
            for i in range(total_n_iters):
                x_batch, y_batch = data_loader(i, x, y)
                np_res += np_hsic(x_batch, y_batch, sigma_x, sigma_y)

            expected_hsic = np_res / total_n_iters
            assert pytest.approx(expected_hsic, abs=tol) == res
```

**Review by vfdev-5 — APPROVED** [review]
Thanks for the PR @kzkadc !
Approving it, currently the failing job is unrelated, PR is good to go once other CI jobs are green.

**Comment by kzkadc** [comment]
@vfdev-5 Thank you for fixing the test!
